### PR TITLE
Add `ModifyCustomSpawnersEvent`

### LINKED ISSUE #1243: Event to add custom spawners to levels
Currently, people are AT'ing ServerLevel's customSpawners field: `private final List<CustomSpawner> customSpawners;`

And then adding their custom spawns into there. These customSpawns are used by Vanilla for Phantoms, Cats, Zombie Seige, and Wandering Villagers. We should provide an events ourselves for adding new customSpawners. Maybe the event takes a BiPredicate of the Level and Level resourcekey and modders return true if this is the right level to add the customSpawner to. And the event also takes a customSpawner's constructor Something::new and then constructs a new instance of the spawner for the dimension that passes the BiPredicate

### PR DESCRIPTION
**Pull request by RaymondBlaze**
Closes #1243.
Introduced the `ServerLevelEvent.CustomSpawners` event for modifying `ServerLevel#customSpawners`.
Unlike the proposal in #1243, this PR uses a simpler implementation, mods subscribing the event may access a fully initialized `ServerLevel` instance and modify the custom spawners list.

### CODE REVIEW

File: src/main/java/net/neoforged/neoforge/event/level/ServerLevelEvent.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+        public CustomSpawners(ServerLevel serverLevel, List<CustomSpawner> customSpawners) {
+            super(serverLevel);
+            this.customSpawners = customSpawners;
+        }
+
+        /**
+         * @return the modifiable custom spawner list.
+         */
+        public List<CustomSpawner> getCustomSpawners() {
```
Comment (Matyrobbrt) [inline src/main/java/net/neoforged/neoforge/event/level/ServerLevelEvent.java]: You should add a method to add directly in the event class.

**Review by Matyrobbrt — DISMISSED** [review]
I'm not exactly keen on introducing yet another event with subclasses prematurely, any reason why the event isn't top level?

**Comment by RaymondBlaze** [comment]
> I'm not exactly keen on introducing yet another event with subclasses prematurely, any reason why the event isn't top level?

I'm not sure about the convention of introducing nested class events, the premature argue seems fair to me, so I'll move it to the top level.
Just for sure, are you suggesting that `ServerLevelEvent` class shouldn't exist at all, or just `CustomSpawnersEvent` should be top level instead of nested?

**Comment by Matyrobbrt** [comment]
It shouldn't exist. Alternatively you can make `LevelEvent` generic.

**Comment by Shadows-of-Fire** [comment]
This event should be a top-level class named either `AddCustomSpawnersEvent` or `RegisterCustomSpawnersEvent` (I leave the choice of name up to you, since we have no concrete guidelines on this feature and events for adding things at server startup are split between `Register` and `Add`).  As mentioned, the class `ServerLevelEvent` need not exist.

Do not make `LevelEvent` generic as part of this change, as it will trigger a large unrelated refactor.

**Comment by RaymondBlaze** [comment]
> This event should be a top-level class named either `AddCustomSpawnersEvent` or `RegisterCustomSpawnersEvent`

I would like to use `ModifyCustomSpawnersEvent` instead, since this event provides full modification to the list (which can be used to remove vanilla custom spawners), while both `Add` and `Register` indicates "add-only".

File: src/main/java/net/neoforged/neoforge/event/level/ModifyCustomSpawnersEvent.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+ * <p>
+ * Subscribe to this event to add/remove {@linkplain CustomSpawner custom spawners} for the level.
+ * <p>
+ * This event is not {@link net.neoforged.bus.api.ICancellableEvent cancellable} and does not have a result.
+ */
+public class ModifyCustomSpawnersEvent extends Event {
+    private final ServerLevel serverLevel;
+    private final List<CustomSpawner> customSpawners;
+
+    public ModifyCustomSpawnersEvent(ServerLevel serverLevel, List<CustomSpawner> customSpawners) {
```
Comment (Spinoscythe) [inline src/main/java/net/neoforged/neoforge/event/level/ModifyCustomSpawnersEvent.java:26]: mark this constructor as `@ApiStatus.Internal`

**Review by Spinoscythe — CHANGES_REQUESTED** [review]

**Review by Spinoscythe — APPROVED** [review]
LGTM

**Review by TelepathicGrunt — APPROVED** [review]

**Pull request merged**